packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -479,6 +497,64 @@ LightningElement.prototype = {
         return vm.shadowRoot;
     },
 
+    get refs(): RefNodes | undefined {
+        const vm = getAssociatedVM(this);
+
+        if (process.env.NODE_ENV !== 'production') {
+            warnIfInvokedDuringConstruction(vm, 'refs');
+        }
+
+        const { refVNodes, hasRefVNodes, cmpTemplate } = vm;
+
+        // If the `cmpTemplate` is null, that means that the template has not been rendered yet. Most likely this occurs
+        // if `this.refs` is called during the `connectedCallback` phase. The DOM elements have not been rendered yet,
+        // so log a warning. Note we also check `isBeingConstructed()` to avoid a double warning (due to
+        // `warnIfInvokedDuringConstruction` above).
+        if (
+            process.env.NODE_ENV !== 'production' &&
+            isNull(cmpTemplate) &&
+            !isBeingConstructed(vm)
+        ) {
+            logError(
+                `this.refs is undefined for ${getComponentTag(
+                    vm
+                )}. This is either because the attached template has no "lwc:ref" directive, or this.refs was ` +
+                    `invoked before renderedCallback(). Use this.refs only when the referenced HTML elements have ` +
+                    `been rendered to the DOM, such as within renderedCallback() or disconnectedCallback().`
+            );
+        }
+
+        // For backwards compatibility with component written before template refs
+        // were introduced, we return undefined if the template has no refs defined
+        // anywhere. This fixes components that may already have a property called `refs`
+        // and are checking it with `if (this.refs)`.
+        // Note it is not sufficient to just check if `refVNodes` is null or empty,
+        // because a template may have `lwc:ref` defined within a falsy `if:true` block.
+        if (!hasRefVNodes) {
+            return;
+        }
+        // For templates that are using `lwc:ref`, if there are no refs currently available
+        // (e.g. refs inside of a falsy `if:true` block), we return an empty object.
+        if (isNull(refVNodes)) {
+            return EMPTY_REFS;
+        }
+
+        // The refNodes can be cached based on the refVNodes, since the refVNodes
+        // are recreated every time the template is rendered.
+        let refs = refsCache.get(refVNodes);
+
+        if (isUndefined(refs)) {
+            refs = create(null);
```

Comment:
Updating this line would allow you to remove the `!` in the following lines. This is because `create` return `any` which can be `undefined`.

```suggestion
            refs = create(null) as RefNodes;
```